### PR TITLE
Test state methods in Concrete environment 

### LINKED ISSUE #34: Test state methods in Concrete environment
> **[ODHack 5.0] Contributors**: Note that we highly recommend that you comment on the issue by introducing yourself and how you plan on tackling the issue! Just saying you want an issue is not helpful.  This will help with assigning tasks, especially if many comment.

# Rationale

Currently, the Concrete environment is poorly tested.

# Implementation

This depends on #33 being implemented. Once that is done, the mock statedb in `concrete/mock` can be moved to/reimplemented within `concrete/api` to have an in-memory statedb for testing. Both execution and gas usage should be tested.

At the moment, the state methods are:

```go
type Environment interface {
	// ...

	StorageLoad(key common.Hash) common.Hash
	StorageStore(key common.Hash, value common.Hash)
	Log(topics []common.Hash, data []byte)

	GetBalance(address common.Address) *uint256.Int
	GetCode(address common.Address) []byte
	GetCodeSize() int
	GetExternalBalance(address common.Address) *uint256.Int
	GetExternalCode(address common.Address) []byte
	GetExternalCodeSize(address common.Address) int
	GetExternalCodeHash(address common.Address) common.Hash
}
```

Note two more methods will be added by #31 .

### PR DESCRIPTION
**Pull request by fishonamos**
Closes #34

### CODE REVIEW

File: concrete/api/environment_test.go [hunk 1]
```diff
@@ -32,6 +32,111 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+func TestEnvironmentStateMethods(t *testing.T) {
```
Comment (therealbytes) [inline concrete/api/environment_test.go]: Can you rename this to TestStateMethods and move it to the bottom of the file?

File: concrete/api/environment_test.go [hunk 2]
```diff
@@ -32,6 +32,111 @@ import (
...
+	r := require.New(t)
+	config := EnvConfig{IsStatic: false, IsTrusted: false}
+	meterGas := true
+
+	env, statedb, _, _ := NewMockEnvironment(config, meterGas)
```
Comment (therealbytes) [inline concrete/api/environment_test.go]: Rebase on main and use the new version of NewMockEnvironment.

File: concrete/api/environment_test.go [hunk 3]
```diff
@@ -32,6 +32,111 @@ import (
...
+	env.contract.Gas = uint64(1e6)
+
+	t.Run("StorageLoad and StorageStore", func(t *testing.T) {
```
Comment (therealbytes) [inline concrete/api/environment_test.go]: Rename this to "Storage".

File: concrete/api/environment_test.go [hunk 4]
```diff
@@ -32,6 +32,111 @@ import (
...
+		storedValue := statedb.GetState(env.Contract().Address, key)
+		r.Equal(value, storedValue)
+	})
+
+	t.Run("Log", func(t *testing.T) {
+		topics := []common.Hash{common.HexToHash("0x01"), common.HexToHash("0x02")}
+		data := []byte("log data")
+
+		env.Log(topics, data)
+
```
Comment (therealbytes) [inline concrete/api/environment_test.go]: Check that the log has been added to the StateDB.

**Review by therealbytes — CHANGES_REQUESTED** [review]

**Review by therealbytes — CHANGES_REQUESTED** [review]
Running the tests results in a panic.

Also, I don't see why the changes to mock statedb and env are necessary. Could you explain?

**Comment by fishonamos** [comment]
> Running the tests results in a panic.
> 
> Also, I don't see why the changes to mock statedb and env are necessary. Could you explain?

The update you made to mock was not compatible  with existing mockStateDB and the interface. and a lot of merge conflicts when i rebased. So I had to manually fix them to fit the existing state. Let me try to do an a mirror implementation on another branch.  Will post a screenshot of the tests status immediately.

**Comment by fishonamos** [comment]
All the tests are passing except GetCodeSize & GetExternalCodeSize. What could be the reason? Can you check please?

I commented them. Dmed you on TG.

**Comment by therealbytes** [comment]
I just merged a PR that should fix it. Try rebasing.

That said, I still don't see why any of the changes in interface_go.go and mock.go are necessary.

You can use the existing mocks like this:

```golang
func TestStateMethods(t *testing.T) {
	r := require.New(t)
	config := EnvConfig{IsStatic: false, IsTrusted: false}
	meterGas := true

	env, statedb, _, _ := NewMockEnvironment(WithConfig(config), WithMeterGas(meterGas))
	env.contract.Gas = uint64(1e6)

	t.Run("Storage", func(t *testing.T) {
		key := common.HexToHash("0x01")
		value := common.HexToHash("0x02")
		env.StorageStore(key, value)
		loadedValue := env.StorageLoad(key)
		r.Equal(value, loadedValue)
		storedValue := statedb.GetState(env.Contract().Address, key)
		r.Equal(value, storedValue)
	})

	t.Run("Log", func(t *testing.T) {
		topics := []common.Hash{common.HexToHash("0x01"), common.HexToHash("0x02")}
		data := []byte("log data")
		env.Log(topics, data)
		log := &types.Log{
			Address: env.Contract().Address,
			Topics:  topics,
			Data:    data,
		}
		logs := statedb.(*state.StateDB).Logs()
		r.Len(logs, 1)
		r.Equal(log, logs[0])
	})

	t.Run("GetCode", func(t *testing.T) {
		code := []byte{0x60, 0x60, 0x60}
		statedb.(*state.StateDB).SetCode(env.Contract().Address, code)
		loadedCode := env.GetCode()
		r.Equal(code, loadedCode)
	})

	t.Run("GetCodeSize", func(t *testing.T) {
		code := []byte{0x60, 0x60, 0x60}
		statedb.(*state.StateDB).SetCode(env.Contract().Address, code)
		loadedCodeSize := env.GetCodeSize()
		r.Equal(len(code), loadedCodeSize)
	})

	t.Run("GetExternalBalance", func(t *testing.T) {
		address := common.HexToAddress("0x12345")
		balance := uint256.NewInt(5000)
		statedb.(*state.StateDB).SetBalance(address, balance)
		loadedBalance := env.GetExternalBalance(address)
		r.Equal(balance, loadedBalance)
	})

	t.Run("GetExternalCode", func(t *testing.T) {
		address := common.HexToAddress("0x12345")
		code := []byte{0x61, 0x61, 0x61, 0x61}
		statedb.(*state.StateDB).SetCode(address, code)
		loadedCode := env.GetExternalCode(address)
		r.Equal(code, loadedCode)
	})

	t.Run("GetExternalCodeSize", func(t *testing.T) {
		address := common.HexToAddress("0x12345")
		code := []byte{0x61, 0x61, 0x61, 0x61}
		statedb.(*state.StateDB).SetCode(address, code)
		loadedCodeSize := env.GetExternalCodeSize(address)
		r.Equal(len(code), loadedCodeSize)
	})

	t.Run("GetExternalCodeHash", func(t *testing.T) {
		address := common.HexToAddress("0x12345")
		code := []byte{0x61, 0x61, 0x61, 0x61}
		statedb.(*state.StateDB).SetCode(address, code)
		loadedCodeHash := env.GetExternalCodeHash(address)
		r.Equal(crypto.Keccak256Hash(code), loadedCodeHash)
	})
}
```

The only things missing here would be checking the gas costs are correct.

**Comment by fishonamos** [comment]
Alright. Checking

**Comment by fishonamos** [comment]
Nice @therealbytes . The update to the env fixed the tests!

**Comment by fishonamos** [comment]
too many conflicts on this branch. Requesting to close in favour of #73 . I will leave it open for now as reference.

**Pull request closed without merging**